### DISCUSSION TITLE
[wish snippet] Add a new `TRAK` axis in an optimal way

### DISCUSSION THREAD

**Discussion by twardoch** [fonttools/fonttools#2564] [Ideas]
Could someone perhaps write a snippet that shows how to add a `TRAK` axis to an existing VF? 

- The user values should go from -1000 to +1000
- At -1000, there should a delta that reduces the advance width of all glyphs by `-head.unitsPerEm`
- At +1000, there should a delta that increases the advance width of all glyphs by `+head.unitsPerEm`
- Can work in just `gvar`-based (TT) fonts
- The `fvar` tag should be `TRAK` and the name should be `Tracking`
- No `fvar` instances

I imagine that this could be done in quite optimized way, but I’m not sure how to approach this. I think it would be generally useful, and it’s something that is best added onto a finished font rather than on the source level.

**Comment by simoncozens** [comment]
```python
from fontTools.ttLib import TTFont
from fontTools.ttLib.tables._f_v_a_r import Axis
import fontTools.ttLib.tables.TupleVariation as tv
import argparse

TupleVariation = tv.TupleVariation

parser = argparse.ArgumentParser(description="Add a TRAK axis")
parser.add_argument("-o", "--output", help="output font file")
parser.add_argument("font", help="font to add the axis to")

args = parser.parse_args()

font = TTFont(args.font)

if "fvar" not in font:
    raise Exception("Font does not have fvar table")

var = font["gvar"].variations
for glyphname, value in var.items():
    glyph = font["glyf"][glyphname]
    numPointsInGlyph = font["gvar"].getNumPoints_(glyph)

    minusValue = TupleVariation(
        {"TRAK": (-1.0, -1.0, 0.0)}, [(0, 0)] * numPointsInGlyph
    )
    minusValue.coordinates[-3] = (-font["head"].unitsPerEm, 0)

    plusValue = TupleVariation({"TRAK": (0.0, 1.0, 1.0)}, [(0, 0)] * numPointsInGlyph)
    plusValue.coordinates[-3] = (font["head"].unitsPerEm, 0)

    value.append(minusValue)
    value.append(plusValue)

# HVAR is horrible and I hate it.
if "HVAR" in font:
    del font["HVAR"]

# Add axis at the end after gvar is decompiled
trak = Axis()
trak.axisTag = "TRAK"
trak.minValue, trak.defaultValue, trak.maxValue = -1000, 0, 1000
trak.axisNameID = font["name"].addName("Tracking")
fvar = font["fvar"]
fvar.axes.append(trak)

# Each named instance needs to add the axis too, even though we're not using it there.
for instance in fvar.instances:
    instance.coordinates["TRAK"] = 0

font.save(args.output)
```

**Comment by benkiel** [comment]
> # HVAR is horrible and I hate it.

That was a good chuckle.

**Comment by twardoch** [comment]
excellent, thank you!

**Comment by twardoch** [comment]
I’ll briefly clarify why this is useful: with instancer, I can this way »freeze« letter-spacing into an instance. And remix with other axes (wght, wdth for example) to simulate opsz :)

**Comment by twardoch** [comment]
The snippet works, the output font works in Samsa & FontView & FontGoggles, but in Safari & Chrome changing any variations makes the glyphs disappear.
